jerry-core/lit/lit-strings.c

```diff
@@ -31,89 +31,53 @@ lit_is_valid_utf8_string (const lit_utf8_byte_t *utf8_buf_p, /**< utf-8 string *
                           lit_utf8_size_t buf_size, /**< string size */
                           bool is_strict) /**< true if surrogate pairs are not allowed */
 {
-  lit_utf8_size_t idx = 0;
+  if (!utf8_buf_p)
+  {
+    return false;
+  }
 
-  bool is_prev_code_point_high_surrogate = false;
-  while (idx < buf_size)
+  if (buf_size == 0)
   {
-    lit_utf8_byte_t c = utf8_buf_p[idx++];
-    if ((c & LIT_UTF8_1_BYTE_MASK) == LIT_UTF8_1_BYTE_MARKER)
-    {
-      is_prev_code_point_high_surrogate = false;
-      continue;
-    }
+    return true;
+  }
 
-    lit_code_point_t code_point = 0;
-    lit_code_point_t min_code_point = 0;
-    lit_utf8_size_t extra_bytes_count;
-    if ((c & LIT_UTF8_2_BYTE_MASK) == LIT_UTF8_2_BYTE_MARKER)
+  lit_utf8_size_t extra_bytes_count;
+
+  for (const unsigned char *idx = (const unsigned char *) utf8_buf_p; (unsigned int) (idx - utf8_buf_p) < buf_size;
+       idx += extra_bytes_count)
+  {
+    if (idx[0] <= 0x7F)
     {
       extra_bytes_count = 1;
-      min_code_point = LIT_UTF8_2_BYTE_CODE_POINT_MIN;
-      code_point = ((uint32_t) (c & LIT_UTF8_LAST_5_BITS_MASK));
+      continue;
     }
-    else if ((c & LIT_UTF8_3_BYTE_MASK) == LIT_UTF8_3_BYTE_MARKER)
+    else if (((0xC2 <= idx[0] && idx[0] <= 0xDF) && (0x80 <= idx[1] && idx[1] <= 0xBF)))
     {
       extra_bytes_count = 2;
-      min_code_point = LIT_UTF8_3_BYTE_CODE_POINT_MIN;
-      code_point = ((uint32_t) (c & LIT_UTF8_LAST_4_BITS_MASK));
+      continue;
     }
-    else if ((c & LIT_UTF8_4_BYTE_MASK) == LIT_UTF8_4_BYTE_MARKER)
+    else if ((idx[0] == 0xE0 && (0xA0 <= idx[1] && idx[1] <= 0xBF) && (0x80 <= idx[2] && idx[2] <= 0xBF))
+              || (((0xE1 <= idx[0] && idx[0] <= 0xEC) || idx[0] == 0xEE || idx[0] == 0xEF)
+              && (0x80 <= idx[1] && idx[1] <= 0xBF) && (0x80 <= idx[2] && idx[2] <= 0xBF))
+              || (idx[0] == 0xED && (0x80 <= idx[1] && idx[1] <= 0x9F) && (0x80 <= idx[2] && idx[2] <= 0xBF))
+              || (idx[0] == 0xED && is_strict && (((0xA0 <= idx[1] && idx[1] <= 0xAF)
+              && (0x80 <= idx[2] && idx[2] <= 0xBF)) || ((0xB0 <= idx[1] && idx[1] <= 0xBF)
+              && (0x80 <= idx[2] && idx[2] <= 0xBF)))))
```

Comment:
This is very hard to understand. I would split it into multiple `if`s. Use lowercase for hex constants.

Very simplified example (use better names and syntax):
```
c = *buf++; // No need to increase the buf-ptr in the loop
if (c < 0x80) continue;
if (c < 0xc2) return false;
if (buf >= end) return false;
n = *buf++;
if ((n & 0xc0) != 0x80) return false;
if (c < 0xe0) continue;
if (buf >= end) return false;
if ((*buf++ & 0xc0) != 0x80) return false;
if (c < 0xf0) {
  // three byte case, check the validity
  if (c == 0xe0 && n < ...) return false;
  // Check surrogates
  continue;
}
// Continue with 4 byte case
if ((*buf++ & 0xc0) != 0x80) return false;
...
```